From 78000228cd8302e3329a243d55c1fffa919b424a Mon Sep 17 00:00:00 2001
From: ran shalit <rshalit@habana.ai>
Date: Sat, 22 May 2021 11:29:46 +0300
Subject: [PATCH] [SW-43026] fw_upd: store closure in boot_info

Store fw_upd closure in boot_info in case of fit
command to close fw_upd module. This information
is preserved in sw reset, and will also
be used by u-boot for closure of fw_upd module
in a later patch.

Change-Id: I81f32bf7c7a84b83187a46f68517c6ba44a7aafa
Signed-off-by: ran shalit <rshalit@habana.ai>
---

diff --git a/common/cmd_module.h b/common/cmd_module.h
index 9faedbf..e9e0ecb 100644
--- a/common/cmd_module.h
+++ b/common/cmd_module.h
@@ -86,6 +86,7 @@
 void module_fwupd_run(void);
 void module_fwupd_run_noret(void);
 void module_fwupd_pcie_disable(bool is_pcie_disabled);
+bool module_fwupd_is_disabled(void);
 
 /* I2C FWUPD */
 void module_i2c_fwupd_setup(void);
diff --git a/gaudi/asic.c b/gaudi/asic.c
index 1ebfa1e..4f70169 100644
--- a/gaudi/asic.c
+++ b/gaudi/asic.c
@@ -30,6 +30,8 @@
 #include <crc32.h>
 #include <pci.h>
 #include <security.h>
+#include <boot_info.h>
+#include <cmd_module.h>
 
 #include "common.h"
 
@@ -251,6 +253,18 @@
 	preboot_boot_info_set_fw_info();
 }
 
+static u32 fwupd_pcie_disable;
+
+void preboot_finalize_fw_upd_closure(void)
+{
+	fwupd_pcie_disable = module_fwupd_is_disabled();
+
+	if (fwupd_pcie_disable) {
+		boot_info_set(HL_BOOT_INFO_FW_UPD_CLOSURE, &fwupd_pcie_disable,
+			      sizeof(fwupd_pcie_disable));
+	}
+}
+
 #define PPBOOT_OFFSET_PRIMARY	0x00000
 #define PPBOOT_OFFSET_SECONDARY	0x20000
 #define PPBOOT_MAX_SIZE		0x20000
diff --git a/include/preboot.h b/include/preboot.h
index 172f9da..8a3e6c3 100644
--- a/include/preboot.h
+++ b/include/preboot.h
@@ -185,6 +185,7 @@
 void preboot_finalize_nic_en(void);
 void preboot_boot_info_set_fw_info(void);
 void preboot_finalize_rma_tables(void);
+void preboot_finalize_fw_upd_closure(void);
 
 bool preboot_is_image_sig_supported(void);
 int preboot_verify_img_sig(const void *fit, size_t img_size,
diff --git a/preboot/fit.c b/preboot/fit.c
index 4e2f52f..6602bae 100644
--- a/preboot/fit.c
+++ b/preboot/fit.c
@@ -967,6 +967,7 @@
 __weak void preboot_finalize_fw_info(void) {}
 __weak void preboot_finalize_rma_tables(void) {}
 __weak void preboot_secondary_cpu_stop(void) {}
+__weak void preboot_finalize_fw_upd_closure(void) {}
 
 void fit_run_image(void)
 {
@@ -984,6 +985,7 @@
 	preboot_finalize_nic_en();
 	preboot_finalize_fw_info();
 	preboot_finalize_rma_tables();
+	preboot_finalize_fw_upd_closure();
 	preboot_finalize_enabled_caps();
 	preboot_secondary_cpu_stop();
 	boot_info_finalize();
diff --git a/preboot/module_fwupd.c b/preboot/module_fwupd.c
index c361fe2..794a72f 100644
--- a/preboot/module_fwupd.c
+++ b/preboot/module_fwupd.c
@@ -364,6 +364,11 @@
 	state.is_disabled = is_pcie_disabled;
 }
 
+bool module_fwupd_is_disabled(void)
+{
+	return state.is_disabled;
+}
+
 static enum hl_upd_sts module_fwupd_map(struct fwupd_state *state,
 					unsigned int size, enum hl_upd_cmd cmd)
 {
diff --git a/preboot/module_lkd.c b/preboot/module_lkd.c
index 4e6e74a..053716d 100644
--- a/preboot/module_lkd.c
+++ b/preboot/module_lkd.c
@@ -550,6 +550,7 @@
 {
 	preboot_finalize_fw_info();
 	preboot_finalize_nic_en();
+	preboot_finalize_fw_upd_closure();
 	boot_info_finalize();
 }
 
diff --git a/preboot/preboot.c b/preboot/preboot.c
index 1352e59..096d28a 100644
--- a/preboot/preboot.c
+++ b/preboot/preboot.c
@@ -309,8 +309,9 @@
 {
 	struct hl_fw_info_compat fw_info_local = {};
 	u32 boot_err = HL_FW_FIT_BOOT_ERR_NONE;
-	u32 data_size;
+	u32 data_size, rst_src;
 	u32 nic_en_local;
+	u32 fw_upd_is_dis = 0;
 
 	if (!boot_info_get(HL_BOOT_INFO_FIT_BOOT_ERROR, &boot_err,
 			   sizeof(boot_err), &data_size)) {
@@ -336,6 +337,17 @@
 			   sizeof(nic_en_local), &data_size))
 		preboot_nic_en(nic_en_local);
 
+	/*
+	 * Preserve pci disable if exist, except for PRSTN, in which it is
+	 * required to clear previous indication.
+	 */
+	rst_src = regmap_readl(mmPSOC_GLOBAL_CONF_RST_SRC);
+	if (rst_src != HL_PRSTN_RST)
+		boot_info_get(HL_BOOT_INFO_FW_UPD_CLOSURE, &fw_upd_is_dis,
+			      sizeof(fw_upd_is_dis), &data_size);
+
+	module_fwupd_pcie_disable(fw_upd_is_dis);
+
 	if (!boot_info_get(HL_BOOT_INFO_FW_INFO, &fw_info_local,
 			   sizeof(fw_info_local), &data_size)) {
 		if (fw_info_local.version == HL_FW_INFO_VERSION)
